src/compute/src/render/persist_source.rs

```diff
@@ -0,0 +1,209 @@
+                }
+            }
+    };
+
+    let mut pinned_stream = Box::pin(async_stream);
+
+    let (timely_stream, token) =
+        mz_storage::source::util::source(scope, "persist_source".to_string(), move |info| {
+            let activator = Arc::new(scope.sync_activator_for(&info.address[..]));
+
+            // There is a bit of a mismatch: listening on a ReadHandle will give us an Antichain<T>
+            // as a frontier in `Progress` messages while a timely source usually only has a single
+            // `Capability` that it can downgrade. We can work around that by using a
+            // `CapabilitySet`, which allows downgrading the contained capabilities to a frontier
+            // but `util::source` gives us both a vanilla `Capability` and a `CapabilitySet` that
+            // we have to keep downgrading because we cannot simply drop the one that we don't
+            // need.
+            let mut current_ts = 0;
+
+            move |cap, cap_set, output| {
+                let waker = futures_util::task::waker_ref(&activator);
+                let mut context = Context::from_waker(&waker);
+
+                while let Poll::Ready(item) = pinned_stream.as_mut().poll_next(&mut context) {
+                    match item {
+                        Some(Ok(event)) => match event {
+                            ListenEvent::Progress(upper) => {
+                                // NOTE(aljoscha): Ideally, we would only get a CapabilitySet, that
+                                // we have to manage. Or the ability to drop the plain Capability
+                                // right at the start.
+                                if let Some(first_element) = upper.first() {
+                                    current_ts = *first_element;
+                                    cap.downgrade(first_element);
+                                }
+                                cap_set.downgrade(upper);
+                            }
+                            ListenEvent::Updates(updates) => {
+                                let cap = cap_set.delayed(&current_ts);
+                                let mut session = output.session(&cap);
+                                for update in updates {
+                                    session.give(Ok(update));
+                                }
+                            }
+                        },
+                        Some(Err::<_, ExternalError>(e)) => {
+                            let cap = cap_set.delayed(&current_ts);
+                            let mut session = output.session(&cap);
+                            session.give(Err((format!("{}", e), current_ts, 1)));
+                        }
+                        None => {
+                            cap_set.downgrade(&[]);
+                            // Returning `Done` here will drop the capabilities.
+                            return SourceStatus::Done;
+                        }
```

Comment:
What does `None` mean in this context? Downgrading the capability to `[]` seems to be something pretty strong, not just e.g. "connection severed" or the like. It seems like maybe `None` when the frontier is not `[]` could/should mean "disconnection error"?

Reply:
It means that the stream (which we're setting up above) is exhausted, but that should never happen because we keep polling for new updates from persist forever in the asyn block above. I'll change this to panic if the stream is ever exhausted.

Reply:
That sounds great; thanks!